### DISCUSSION TITLE
async functions, "require-await" and "no-return-await"

### DISCUSSION THREAD

**Discussion by ziadkh0** [eslint/eslint#14567] [Q&A] [answered]
Hello everyone,

I'm not sure if my use case should be handled by the rules mentioned in the title, so I thought I'd start a discussion instead of an issue..

I have a function defined like this:
```js
async function foo() {
  return window.bar || await baz();
}
```
As is, this will trigger the "no-return-await" rule [demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tcmV0dXJuLWF3YWl0OiAnZXJyb3InKi9cbmFzeW5jIGZ1bmN0aW9uIGZvbygpIHtcbiAgcmV0dXJuIHdpbmRvdy5iYXIgfHwgYXdhaXQgYmF6KCk7XG59Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMiwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==).

And if I change it to:
```js
async function foo() {
  return window.bar || baz();
}
```
It will trigger the "require-await" rule [demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgcmVxdWlyZS1hd2FpdDogJ2Vycm9yJyovXG5hc3luYyBmdW5jdGlvbiBmb28oKSB7XG4gIHJldHVybiB3aW5kb3cuYmFyIHx8IGJheigpO1xufSIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTIsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)

So my question, is there a better way to write it that satisfies the lint rules and not end up less readable?
Or should the rules be changed to allow such a use case?

Thanks in advance

**Comment by nzakas** [comment, marked as answer]
The complexity here is that your return statement doesn’t always return a promise, so there isn’t a way to satisfy both rules. You need to disable one of them.

**Reply by ziadkh0** [reply]
Thank you for the answer :smile: 
After thinking more about this, I feel that disabling "require-await" is the way to go.

Because:
1. "no-return-await" has a valid point (we shouldn't double wrap `baz()`'s answer), and
2. "require-await" wants to avoid programmer error (ie. marking a sync function async) which isn't the case here.
